### ISSUE TITLE
Create Travis CI example of using MXE binary distribution

### ISSUE THREAD

**Issue by starius** [mxe/mxe#962]
Current repo for this task: https://github.com/starius/mxe-travis-example (empty)
Can we move it to github.com/mxe/mxe-travis-example when it's finished?

**Comment by tonytheodore** [comment]
That is a very good idea! Building MXE on Travis isn't feasible, but actual demonstrations of using it would be great. I wouldn't call it "example" though, pick a real project (one of yours or projects using mxe preferably) and name it accordingly.

No objections here to moving it into `/mxe`. The only objections I might have would be around "toy" projects or "unofficial" builds - though I have no solid definition of those.

**Comment by tonytheodore** [comment]
Thinking about #924, I wonder if we could do something like:
1. get list of packages affected by pull request
2. get list of deps (`make show-upstream-deps-'<pkgs>'`)
3. `make --touch <deps>`
4. install `<deps>` from packages
5. build pull request

**Comment by starius** [comment]
Good idea!

Though binary packages have to be rebuilt with less delay.
I have built commit 19d0acd474e39fbc2c2f7f136f9abaafdfcdc337:
https://gist.github.com/starius/0688b469a72d1010351a
After #1012 is fixed, I'm going to rebuild MXE and replace .deb files on redjohn APT.

**Comment by tonytheodore** [comment]
#1014 fixes #1012 and has no impact on any other packages here.

**Comment by starius** [comment]
I'm building 702b214c8316c347fb2ee75e07d36af70a7e0c4e with applied 9bd2349eebe5489a4712bb0678cdf4eb4a42f276.

**Comment by starius** [comment]
The build succeeded.
See [the log](https://gist.github.com/starius/9191afe744b83821674a).

**Comment by starius** [comment]
The build of 18237df9b232da59a76d7342d1dc7bc7d607486e is ready, no packages are broken.
Please see [the log](https://gist.github.com/starius/2592dca4439e1d53587f). If everything is OK, i'll deploy it on mxe.redjohn.tk.

**Comment by tonytheodore** [comment]
I've created a new issue for qwt download #1048, it can happen when packages share the same file. Some of the other were subsequently fixed in https://github.com/mxe/mxe/compare/18237df...b2e57a7 and https://github.com/mxe/mxe/compare/9df3364...0a59413.

Looks good to go, thanks!

**Comment by tonytheodore** [comment]
> deploy it on mxe.redjohn.tk

Should we create some sort of redirect on `?.mxe.cc` to document as the "official" repos?

**Comment by tonytheodore** [comment]
Is it possible to deploy a single package like `mxe-conf` if the cmake toolchain file is updated (say a single line change like #1023)?

**Comment by starius** [comment]
> Is it possible to deploy a single package like mxe-conf if the cmake toolchain file is updated (say a single line change like #1023)?

I would prefer to rebuild all packages from scratch to make sure that deployed files can be reproduced from a particular commit.

> Should we create some sort of redirect on ?.mxe.cc to document as the "official" repos?

Yes, you can create a subdomain e.g., pkg.mxe.cc -> 62.210.92.11. I'll add it to my nginx.conf as a mirror. URL with instructions in `tools/build-pkg.lua` should be updated as well.

**Comment by tonytheodore** [comment]
Great! @vog, could you create the above subdomain and this one also?

mirror.mxe.cc -> 104.131.71.203 2604:a880:800:10::1

**Comment by vog** [comment]
Done. Note that the changes may take up to 24 hours to be effective.

In total, the domain `mxe.cc` is now configured as follows:

    @                        IN A       192.30.252.154
    @                        IN A       192.30.252.153
    mirror                   IN A       104.131.71.203
    mirror                   IN AAAA    2604:a880:800:10::1
    pkg                      IN A       62.210.92.11
    pkg                      IN AAAA    2001:bc8:338c::1
    www                      IN CNAME   mxe.github.io.

**Comment by starius** [comment]
http://pkg.mxe.cc/ works for building NPG-explorer for Windows:
https://travis-ci.org/npge/npge/builds/96141354

**Comment by TimothyGu** [comment]
@vog, you might want to change `www` to be an A record for `mxe.cc`. It's currently redirecting to mxe.cc anyway, and GitHub pages discourages the use of CNAME IIRC.

**Comment by tonytheodore** [comment]
`ssh` to mirror.mxe.cc working also.

**Comment by vog** [comment]
@TimothyGu: Thanks for the hint. The domain `mxe.cc` is now configured as follows:

    @                        IN A       192.30.252.154
    @                        IN A       192.30.252.153
    mirror                   IN A       104.131.71.203
    mirror                   IN AAAA    2604:a880:800:10::1
    pkg                      IN A       62.210.92.11
    pkg                      IN AAAA    2001:bc8:338c::1
    www                      IN CNAME   mxe.cc.

**Comment by TimothyGu** [comment]
:+1:

**Comment by starius** [comment]
I have tried to deploy MXE packages built from 18237df9b232da59a76d7342d1dc7bc7d607486e

Two breaking problems were fixed in pull requests:

  * #1058 build-pkg: do not provide empty dependency list
  * #1056 build-pkg: build empty dependencies of non-empty

Can you review and merge them?

I have created a project for blank APT directory to fill it from MXE directory with build packages: https://github.com/starius/mxe-apt
Now all I have to do to deploy new packages is cloning that repo and type `./make-apt.sh /usr/lib/mxe` and change the symlink to `repos` dir.
Can I move that project to MXE org on GitHub?

Further steps are automation of testing of the APT:

  * deploy new packages on some testing APT like `pkg.mxe.cc/repos-test`.
  * run Travis CI build which uses that testing APT.
  * if the Travis CI build is OK, change the symlink `pkg.mxe.cc/repos` (production APT).
  * send some reports to the mailing list and add tag `build-YYYY-MM-DD` to MXE git.

**Comment by tonytheodore** [comment]
> Can you review and merge them?

Done.

> Can I move that project to MXE org on GitHub?

Certainly, it looks ready to go!

**Comment by starius** [comment]
>> Can I move that project to MXE org on GitHub?

> Certainly, it looks ready to go!

Done: https://github.com/mxe/mxe-apt

**Comment by starius** [comment]
The build succeeded, but the packages are not installable because of a problem fixed in #1071

Finally two projects using Travis CI and MXE were built successfully:

  * https://travis-ci.org/npge/npge/builds/96505530
  * https://travis-ci.org/starius/battleship/builds/96503948

I want to rebuild it from scratch with #1069, #1070, #1071 and #1072. Can you review and merge them?

By the way the second project is small enough being a real project, not a toy. I think it can serve as MXE-in-Travis example.

**Comment by tonytheodore** [comment]
> Can you review and merge them?

All look good to me, thanks. I'm not sure `mxe-requirements` (#1071) need to be the "full" requirements, though everything is reasonable for any development setup and it's not worth optimising the CI case.

> By the way the second project is small enough being a real project, not a toy. I think it can serve as MXE-in-Travis example.

+1. `cmake`, `qt`, and a quick build are perfect!

**Comment by starius** [comment]
MXE target and deploy to GitHub releases were [merged](https://github.com/zer0main/battleship/pull/3) into battleship upstream. The project was built successfully and files `battleship` (for Linux) and `battleship.exe` (for Windows, built with MXE) were uploaded to [GitHub releases](https://github.com/zer0main/battleship/releases).

**Comment by tonytheodore** [comment]
Nice! Seems like living in the future.
